File: bff-api/internal/services/wordcloud_service.go
package services

import (
	"encoding/csv"
	"log"
	"os"
	"regexp"
	"sort"
	"strings"

	model "github.com/maxiking445/bff-api/internal/models"
)

func GetWordCount(files []string, limit int) ([]model.WordCount, error) {
	wordCounts := make(map[string]int)

	for _, f := range files {
		if err := countWordsInFile(f, wordCounts, true); err != nil {
			log.Printf("error reading %s: %v", f, err)
		}
	}

	// Map -> Slice and sort
	results := make([]model.WordCount, 0, len(wordCounts))
	for w, c := range wordCounts {
		results = append(results, model.WordCount{Word: w, Count: c})
	}
	sort.Slice(results, func(i, j int) bool {
		if results[i].Count == results[j].Count {
			return results[i].Word < results[j].Word
		}
		return results[i].Count > results[j].Count
	})

	// shorten with limit
	if limit < len(results) {
		results = results[:limit]
	}
	return results, nil
}

func countWordsInFile(path string, wordCounts map[string]int, showOnlyUserWords bool) error {
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()

	r := csv.NewReader(f)
	r.ReuseRecord = true

	records, err := r.ReadAll()
	if err != nil {
		return err
	}
	if len(records) == 0 {
		return nil
	}

	header := records[0]
	bodyIdx := -1
	for i, col := range header {
		if col == "body" {
			bodyIdx = i
			break
		}
	}
	if bodyIdx == -1 {
		return nil //
	}

	// Regex, to limit to only words
	wordRegexp := regexp.MustCompile(`[A-Za-z0-9ÄÖÜäöüß]+`)

	for _, row := range records[1:] {
		if bodyIdx >= len(row) {
			continue
		}
		text := row[bodyIdx]
		text = strings.ToLower(text)

		if strings.Contains(text, "image") {
			continue
		}

		if showOnlyUserWords {
			if strings.Contains(row[5], "READ") {
				continue
			}
		}

		urlRegexp := regexp.MustCompile(`^(https?://|www\.)\S+$`)
		if urlRegexp.MatchString(text) {
			continue
		}

		words := wordRegexp.FindAllString(text, -1)
		for _, w := range words {
			if w == "" {
				continue
			}
			wordCounts[w]++
		}
	}

	return nil
}

File: analyzer-frontend/src/service/WordCloudService.ts
import Papa from "papaparse";
import { listFiles, getZipFile } from "./FileService";
import { ModelsWordCount } from "@/models";

const wordRegexp = /[A-Za-z0-9ÄÖÜäöüß]+/g;
const urlRegexp = /^(https?:\/\/|www\.)\S+$/;
const messageFileRegexp = /^message_.*\.csv$/;

function countWordsInContent(
  content: string,
  wordCounts: Map<string, number>,
  showOnlyUserWords: boolean,
): void {
  const { data: records } = Papa.parse<string[]>(content, {
    skipEmptyLines: true,
  });

  if (records.length === 0) return;

  const header = records[0];
  const bodyIdx = header.indexOf("body");
  if (bodyIdx === -1) return;

  for (const row of records.slice(1)) {
    if (bodyIdx >= row.length) continue;

    const text = row[bodyIdx].toLowerCase();

    if (text.includes("image")) continue;

    if (showOnlyUserWords && row[5]?.includes("READ")) continue;

    if (urlRegexp.test(text)) continue;

    const words = text.match(wordRegexp) ?? [];
    for (const word of words) {
      if (word) {
        wordCounts.set(word, (wordCounts.get(word) ?? 0) + 1);
      }
    }
  }
}

export async function getWordCount(limit: number): Promise<ModelsWordCount[]> {
  const wordCounts = new Map<string, number>();

  const matchingFiles = listFiles().filter((f) => messageFileRegexp.test(f));

  for (const filename of matchingFiles) {
    try {
      const entry = getZipFile(filename);
      if (!entry) continue;

      const content = await entry.content.text();
      countWordsInContent(content, wordCounts, true);
    } catch (error) {
      console.error(`Error reading ${filename}:`, error);
    }
  }

  const results: ModelsWordCount[] = Array.from(wordCounts.entries()).map(
    ([word, count]) => ({ word, count }),
  );

  results.sort((a, b) => {
    if (a.count === b.count) return (a.word ?? "").localeCompare(b.word ?? "");
    return (b.count ?? 0) - (a.count ?? 0);
  });

  return results.slice(0, limit);
}